### DISCUSSION TITLE
With TypeScript 5.7 getData returns ArrayBufferLike instead of ArrayBuffer

### DISCUSSION THREAD

**Discussion by Scarabol** [gildas-lormeau/zip.js#584] [Q&A]
Thank you all for this great library! I appreciate using it in [my project](https://github.com/Scarabol/rock-raiders-web). I just tried to update my TypeScript version from `5.6.3` to `5.7.2`.

Starting with Typescript 5.7 typed arrays become generic. See:
- https://devblogs.microsoft.com/typescript/announcing-typescript-5-7/#support-for---target-es2024-and---lib-es2024
- https://github.com/microsoft/TypeScript/pull/59417

This affects for example the type definition of [Uint8ArrayWriter](https://github.com/gildas-lormeau/zip.js/blob/master/index.d.ts#L625)

Previously with TypeScript 5.6.3 the following code typed `data` as `ArrayBuffer`, but starting with TypeScript > 5.7 `data` has type `ArrayBufferLike`.
```
// TypeScript < 5.7
const data: ArrayBuffer = await entry.getData(new Uint8ArrayWriter());
// TypeScript >= 5.7
const data: ArrayBufferLike = await entry.getData(new Uint8ArrayWriter());
```

My tsconfig compiler target is and has been set to ES2021.

I'm not sure how this is solved properly. My guess would be to add generic type explicitly to all `TypedArrays` usages, since they use Uint8Array<ArrayBuffer> ([Example](https://github.com/gildas-lormeau/zip.js/blob/master/lib/core/io.js#L526)) actually. Effectively changing [type definition here](https://github.com/gildas-lormeau/zip.js/blob/master/index.d.ts#L625) into this:
```
export class Uint8ArrayWriter extends Writer<Uint8Array<ArrayBuffer>> { }
```

What do you think?

**Comment by gildas-lormeau** [comment]
Thank for the info. Actually, I don't use TS for development and I was not aware of this issue. I wonder how to deal with that. I guess somehow I would need to publish 2 `index.d.ts` files?

**Comment by UnsungHero97** [comment]
I don't think it's necessary to publish 2 `index.d.ts` files. I don't see any uses of `SharedArrayBuffer` in the codebase so I think @Scarabol's suggestion is probably the way to do it, here's a pull request for that: https://github.com/gildas-lormeau/zip.js/pull/561

However, I'm not totally sure how this impacts TypeScript versions below 5.7.

**Comment by gildas-lormeau** [comment]
I accepted the PR, we'll see if it's a problem :p

**Comment by gildas-lormeau** [comment]
Unfortunately, I've reverted the commit because Angular 18 is depending on a older version of TypeScript, see #563.

**Comment by UnsungHero97** [comment]
I'll try to spend some time this week to figure this out.

**Comment by gildas-lormeau** [comment]
Thanks! I wonder if waiting would be the easiest solution. It seems that this problem doesn't really bother developers who use TS 5.7+ today.

**Comment by gildas-lormeau** [comment]
Actually, I'm probably wrong. For the record, ChatGPT suggests these fixes:
 1. replace `Uint8Array` with `Uint8Array<ArrayBufferLike>`, or
 2. add a new `CompatibleUint8Array` type to handle the old and new syntax, i.e. `type CompatibleUint8Array = Uint8Array extends Uint8Array<infer T> ? Uint8Array<T> : Uint8Array;`

I have some doubts regarding the first suggestion. The second one is interesting, I'll try to test it.

EDIT:

None of these suggestions actually work in https://www.typescriptlang.org/play/. However ChatGPT is now saying me that I shouldn't need to change anything in fact...

EDIT 2:

The solution below seems to work.
```ts
type CompatibleUint8Array = 
  Uint8Array extends { new (...args: any[]): infer U } 
    ? U 
    : Uint8Array;
```

EDIT 3:
Indeed, the code below does not trigger any error.

```ts
type CompatibleUint8Array = 
  Uint8Array extends { new (...args: any[]): infer U } 
    ? U 
    : Uint8Array;

function test(val: CompatibleUint8Array): void {
  console.log(val);
}
```

- in TS < 5.7
```ts
const val:Uint8Array = new Uint8Array([1, 2, 3]);
test(val);
```

- in TS 5.7+
```ts
const val:Uint8Array<ArrayBufferLike> = new Uint8Array([1, 2, 3]);
test(val);
```

**Comment by gildas-lormeau** [comment]
This does not work actually. 

The code below should not trigger an error in TS 5.7+ but it does.
```ts
type CompatibleUint8Array = 
  Uint8Array extends { new (...args: any[] ): infer U } 
    ? U 
    : Uint16Array;

function test(val: CompatibleUint8Array): void {
  console.log(val);
}

const val:Uint8Array = new Uint8Array([1, 2, 3]);
test(val);
```

**Comment by gildas-lormeau** [comment]
I'm wondering if it's really an issue. I mean that the code below runs fine in the TS playground with TS 5.7+. @Scarabol, any thoughts?

```ts
function test(val: Uint8Array): void {
  console.log(val);
}

const val:Uint8Array = new Uint8Array([1, 2, 3]);
test(val);

const val2:Uint8Array<ArrayBufferLike> = new Uint8Array([1, 2, 3]);
test(val2);
```

**Comment by Scarabol** [comment]
> This does not work actually.
> 
> The code below should not trigger an error in TS 5.7+ but it does.
> 
> type CompatibleUint8Array = 
>   Uint8Array extends { new (...args: any[] ): infer U } 
>     ? U 
>     : Uint16Array;
> 
> function test(val: CompatibleUint8Array): void {
>   console.log(val);
> }
> 
> const val:Uint8Array = new Uint8Array([1, 2, 3]);
> test(val);

Why did you use `Uint16Array` instead of `Uint8Array`? I did some research and I think a conditional type is the way to go here.

**Comment by Scarabol** [comment]
> I'm wondering if it's really an issue. I mean that the code below runs fine in the TS playground with TS 5.7+. [@Scarabol](https://github.com/Scarabol), any thoughts?
> 
> function test(val: Uint8Array): void {
>   console.log(val);
> }
> 
> const val:Uint8Array = new Uint8Array([1, 2, 3]);
> test(val);
> 
> const val2:Uint8Array<ArrayBufferLike> = new Uint8Array([1, 2, 3]);
> test(val2);

First of I don't consider this an actual bug, the library still works great for me, thanks!

The issue appears when using `.getData(new Uint8ArrayWriter())` in TypeScript 5.7+ The code returns an `Uint8Array<ArrayBufferLike>` instead of `Uint8Array<ArrayBuffer>` like in previous versions. If one wants to use the buffer, a cast or type assertion is needed to get `ArrayBuffer`.

Edit:
An additional `getBuffer(): ArrayBuffer` method in parallel to the existing `getData()` method, could be helpful. Unfortunately I don't know the userbase here and can not rate the value of an additional method. From my POV you want to proceed with the `ArrayBuffer` anyway and don't actually need the `Uint8Array`.

**Comment by gildas-lormeau** [comment]
I seem to have made a fundamental error in using `Uint16Array`. I have not idea why I did this. It's probably a terrible lapse of attention on my part.


Edit: I'm remembering now why I used and UInt16Array. This was done on purpose in order to verify if TS would use it or not. However, if I remember corectlu, this testing approach was incorrect but I do not remember why.

Anyway, I agree that adding a `getBuffer` method could indeed be interesting for the reason you mentioned, I don't think it would be too complicated to implement since the code relies on Web Streams internally. I'll take a look at it.

**Comment by gildas-lormeau** [comment]
@Scarabol The method to extract data as `ArrayBuffer` instances has been added in the new version [`v2.7.65`](https://github.com/gildas-lormeau/zip.js/releases/tag/v2.7.65)

**Comment by gildas-lormeau** [comment]
I'm also moving the issue in the discussion tab since it's not a bug a zip.js.
